text/0002-serializable-types.md

```diff
@@ -0,0 +1,413 @@
+- Serialize `AngularVelocity` into the stream
+
+Unlike [Array](#example-array) example above, we do not use `BitSerializer.IsReading` flag to change serialization logic but the value of a serialized flag itself. If `SyncVelocity` flag is set to `true`, both `LinearVelocity` and `AngularVelocity` will also be serialized into the stream — otherwise when it is set to `false`, we will leave `LinearVelocity` and `AngularVelocity` with default values.
+
+#### Recursive Nested Serialization
+
+`BitSerializer` implements `void Serialize<T>(ref T value) where T : INetworkSerializable` method which allows for recursive nested serialization.
+
+Let's have a look at the example below:
+
+```cs
+public struct MyStructA : INetworkSerializable
+{
+    public Vector3 Position;
+    public Quaternion Rotation;
+
+    public void NetworkSerialize(BitSerializer serializer)
+    {
+        serializer.Serialize(ref Position);
+        serializer.Serialize(ref Rotation);
+    }
+}
+
+public struct MyStructB : INetworkSerializable
+{
+    public int SomeNumber;
+    public string SomeText;
+    public MyStructA StructA;
+    
+    public void NetworkSerialize(BitSerializer serializer)
+    {
+        serializer.Serialize(ref SomeNumber);
+        serializer.Serialize(ref SomeText);
+        serializer.Serialize(ref StructA);
+    }
+}
+```
+
+If we were to serialize `MyStructA` alone, it would serialize `Position` and `Rotation` into the stream via `BitSerializer`.
+
+However, if we were to serialize `MyStructB`, it would serialize `SomeNumber` and `SomeText` into the stream, then serialize `StructA` by calling `MyStructA`'s `void NetworkSerialize(BitSerializer)` method which serializes `Position` and `Rotation` into the same stream.
+
+**Note:** Technically, there is no hard-limit on how many `INetworkSerializable` fields you can serialize down the tree hierachy but in practice, there are some memory and bandwidth boundaries you'll need to watch out for.
+
+**Pro-tip:** You can [conditionally serialize](#conditional-serialization) in [recursive nested serialization](#recursive-nested-serialization) scenario and make use of both features! :)
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## INetworkSerializable
+
+`INetworkSerializable` interface will not enforce anything other than `NetworkSerialize(BitSerializer)` method.
+
+```cs
+interface INetworkSerializable
+{
+    void NetworkSerialize(BitSerializer serializer);
+}
+```
+
+## BitSerializer
+
+`BitSerializer` is the main aggregator that implements serialization code for built-in supported types and custom types with `INetworkSerializable` interface.
+
+```cs
+// note: pseudo-code
+class BitSerializer
+{
+    // for reading
+    BitSerializer(BitReader reader) { /* ... */ }
+
+    // for writing
+    BitSerializer(BitWriter writer) { /* ... */ }
+
+    bool IsReading { get; }
+
+    // for common types
+    void Serialize(ref int value)
```

Comment:
Have you thought at all about how bit packing might work with this interface? Is that just up to the user before sending to the serializer?

Reply:
`BitSerializer` is a flip-flop container that holds `BitReader` and `BitWriter` internally. Bit-packing and other serialization optimization techniques should be implemented over there (and there are some packed stuff already).